### PR TITLE
Issue:13 Make a deployment accessible using a public URL

### PR DESCRIPTION
**Pull request by Shraddhak22**
Changes:
Implemented service creation and route creation integrated with kodo deploy

Co-Author - Jaideep Rao
Screenshot from 2020-06-18 17-09-52

### CODE REVIEW

**Review by DhanashriPalodkar — COMMENTED** [review]
you can create deploy_test.go and do unit testing if possible. otherwise I am happy with the code. I will pull and try running locally

**Comment by sbose78** [comment]
Shraddha, could you confirm if the route URL works?

File: pkg/kodo/cmd/deploy.go [hunk 1]
```diff
@@ -67,6 +72,81 @@ func Deploy(deployVar *DeploymentVariables, envVar *EnvironmentVariables) error
...
+				},
+			},
+			Selector: map[string]string{
+				"app": "app-image-deployment",
+			},
+		},
+	}
+	_, serviceError := client.CoreV1().Services(envVar.Namespace).Create(context.TODO(), svc, metav1.CreateOptions{})
+	if serviceError == nil {
+		fmt.Printf("\nService created")
```
Comment (ishitasequeira) [inline pkg/kodo/cmd/deploy.go:107]: You can use fmt.Errorf("Service created")

**Comment by jaideepr97** [comment]
The route url is created, but doesn't point to anything since its just a deployment not an actual image

**Comment by Shraddhak22** [comment]
Command to run - 
deploy --token="token"--server="server" --image=openshift/hello-openshift:latest --replicas="no of replicas" --port="port number" --namespace="namespace"

**Comment by sbose78** [comment]
As far as I remember, `openshift/hello-openshift:latest` is an actual image which if deployed and exposed using a service+route, should be reachable over http.

**Comment by ranakan19** [comment]
I was able to run this and deploy an image to cluster, Looks good to me.

**Comment by ranakan19** [comment]
> As far as I remember, `openshift/hello-openshift:latest` is an actual image which if deployed and exposed using a service+route, should be reachable over http.

I was to get a route over http and the webpage said "Hello Openshift", is this what you are referring to?

**Comment by ranakan19** [comment]
> > As far as I remember, `openshift/hello-openshift:latest` is an actual image which if deployed and exposed using a service+route, should be reachable over http.
> 
> I was to get a route over http and the webpage said "Hello Openshift", is this what you are referring to?

http://myroute-prtest.apps.dev-svc-4.5-061813.devcluster.openshift.com/ is the route i got

**Review by ranakan19 — APPROVED** [review]
LGTM

**Comment by sbose78** [comment]
Yes that's what I was referring to! Thank you :)

File: main.go [hunk 2]
```diff
@@ -61,5 +61,7 @@ var deployCommand = &cobra.Command{
 	Run: func(cm *cobra.Command, args []string) {
 		fmt.Println("Creating image deployment")
 		cmd.Deploy(deployVar, envVar)
+		serviceObj, _ := cmd.Service(deployVar, envVar)
```
Comment (sbose78) [inline main.go:69]: Shouldn't ignore the error ;)

File: main.go [hunk 3]
```diff
@@ -61,5 +61,7 @@ var deployCommand = &cobra.Command{
...
+		cmd.Route(deployVar, envVar, serviceObj)
```
Comment (sbose78) [inline main.go:70]: Shouldn't ignore the error :)

File: pkg/kodo/cmd/deploy.go [hunk 4]
```diff
@@ -67,6 +72,81 @@ func Deploy(deployVar *DeploymentVariables, envVar *EnvironmentVariables) error
...
+	_, clientError := newOpenShiftClient(envVar)
+
+	if clientError != nil {
+		log.Fatal(clientError)
+		return clientError
+	}
+
+	routeObj := &routev1.Route{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "myroute",
```
Comment (sbose78) [inline pkg/kodo/cmd/deploy.go:126]: To avoid naming conflicts, could we generate a unique name using the name of the image we are deploying?

We could do this across the project in a different PR. Could you track a separate issue for this?

**Review by sbose78 — CHANGES_REQUESTED** [review]

File: pkg/kodo/cmd/deploy.go (hunk 4 above)
Comment (sbose78) [reply pkg/kodo/cmd/deploy.go:126]: https://github.com/cli-playground/kodo/issues/18 Assigned to @jaideepr97

File: main.go [hunk 5]
```diff
@@ -39,27 +40,33 @@ func init() {
...
 
 func main() {
-
 	rcommand.Execute()
 }
 
-var listCommand = &cobra.Command{
-	Use: "list",
+var countCommand = &cobra.Command{
+	Use:   "count",
```
Comment (sbose78) [inline main.go:52]: This change isn't relevant to this PR, please remove this.
We should be renaming the command in a different branch/PR.

File: main.go [hunk 6]
```diff
@@ -23,7 +23,8 @@ var versionCommand = &cobra.Command{
 
 func init() {
 	rcommand.AddCommand(versionCommand)
-	rcommand.AddCommand(listCommand)
+	rcommand.AddCommand(countCommand)
```
Comment (sbose78) [inline main.go:26]: This change is unrelated to this PR.
Renaming should be done in a different PR.

**Comment by Shraddhak22** [comment]
All the changes suggested for this PR, has been covered in,
https://github.com/cli-playground/kodo/pull/20.
Adding @jaideepr97  in thread.
